frame/staking/src/voter_bags/mod.rs

```diff
@@ -0,0 +1,513 @@
+/// While the users within any particular bag are sorted in an entirely arbitrary order, the overall
+/// stake decreases as successive bags are reached. This means that it is valid to truncate
+/// iteration at any desired point; only those voters in the lowest bag (who are known to have
+/// relatively little power to affect the outcome) can be excluded. This satisfies both the desire
+/// for fairness and the requirement for efficiency.
+pub struct VoterList<T: Config>(PhantomData<T>);
+
+impl<T: Config> VoterList<T> {
+	/// Remove all data associated with the voter list from storage.
+	pub fn clear() {
+		crate::VoterCount::<T>::kill();
+		crate::VoterBagFor::<T>::remove_all(None);
+		crate::VoterBags::<T>::remove_all(None);
+		crate::VoterNodes::<T>::remove_all(None);
+	}
+
+	/// Regenerate voter data from the `Nominators` and `Validators` storage items.
+	///
+	/// This is expensive and should only ever be performed during a migration, never during
+	/// consensus.
+	///
+	/// Returns the number of voters migrated.
+	pub fn regenerate() -> u32 {
+		Self::clear();
+
+		let nominators_iter = Nominators::<T>::iter().map(|(id, _)| Voter::nominator(id));
+		let validators_iter = Validators::<T>::iter().map(|(id, _)| Voter::validator(id));
+		let weight_of = Pallet::<T>::weight_of_fn();
+
+		Self::insert_many(
+			nominators_iter.chain(validators_iter),
+			weight_of,
+		)
+	}
+
+	/// Decode the length of the voter list.
+	pub fn decode_len() -> Option<usize> {
+		crate::VoterCount::<T>::try_get().ok().map(|n| n.saturated_into())
+	}
+
+	/// Iterate over all nodes in all bags in the voter list.
+	///
+	/// Note that this exhaustively attempts to try all possible bag indices. Full iteration can be
+	/// expensive; it's recommended to limit the number of items with `.take(n)`.
+	pub fn iter() -> impl Iterator<Item = Node<T>> {
+		(0..=BagIdx::MAX).filter_map(|bag_idx| Bag::get(bag_idx)).flat_map(|bag| bag.iter())
+	}
+
+	/// Insert a new voter into the appropriate bag in the voter list.
+	///
+	/// If the voter is already present in the list, their type will be updated.
+	/// That case is cheaper than inserting a new voter.
+	pub fn insert_as(account_id: &AccountIdOf<T>, voter_type: VoterType) {
+		// if this is an update operation we can complete this easily and cheaply
+		if !Node::<T>::update_voter_type_for(account_id, voter_type) {
+			// otherwise, we need to insert from scratch
+			let weight_of = Pallet::<T>::weight_of_fn();
+			let voter = Voter { id: account_id.clone(), voter_type };
+			Self::insert(voter, weight_of);
+		}
+	}
+
+	/// Insert a new voter into the appropriate bag in the voter list.
+	pub fn insert(voter: VoterOf<T>, weight_of: impl Fn(&T::AccountId) -> VoteWeight) {
+		Self::insert_many(sp_std::iter::once(voter), weight_of);
+	}
+
+	/// Insert several voters into the appropriate bags in the voter list.
+	///
+	/// This is more efficient than repeated calls to `Self::insert`.
```

Comment:
since you are a fan of microoptimization :D probably having a separate `fn insert` that does not initialize `counter` and `bags` is even _more_ efficient, rather than using `iter::once`.

Reply:
I mean, you're not wrong, but the cost of creating a temporary, single-item `BTreeMap` and `u32` should be undetectably small compared to the storage costs of insertion.